### ISSUE TITLE
Pixelated STEM backend processing library

### ISSUE THREAD

**Issue by magnunor** [hyperspy/hyperspy#2187]
There are currently several Python packages for working with large pixelated STEM (4D STEM) datasets: `pixStem`, `pyXem`, `py4DSTEM`, `fpd`, `LiberTEM`... This leads some duplication of work, which is not ideal, as discussed in https://github.com/hyperspy/hyperspy/pull/2180. I feel that the most immediate way of reducing this duplication is trying to combine some of the generic "backend" processing functions in one common data processing library. Essentially for working with data with two signal-dimensions.

An example of such a function is radial integration/averaging, which can be implemented easily, but can also quickly increase in complexity if the centre position is different for each diffraction pattern.

So I think ideally such a "backend processing" library should only depend on the more common scientific python stack, so potentially other non-`HyperSpy` based projects could also use them. I'm tempted to also include `dask` in this, since these datasets can easily be terabyte size.

I've already partly implemented this in a separate sub-module in `pixStem`, but I haven't had the time to spin this off into its own library, due to time constraints... https://gitlab.com/pixstem/pixstem/blob/master/pixstem/dask_tools.py

Does this sound like a good idea? Which functionalities are most relevant? Center of mass? Radial integration? Peak finding? Template matching?

Further down the line, combining/merging of the "frontend" functionality could also be a possibility. But I think starting with the "backend" is more appropriate, as it can be done much more rapidly, involves less bikeshedding, and should benefit people who don't use HyperSpy as a their UI.  

(I already made a post about this on the `pyXem` issue tracker (https://github.com/pyxem/pyxem/issues/184) last year, but I haven't had time to follow it up yet. And posting it here now due to the renewed discussion in https://github.com/hyperspy/hyperspy/pull/2180)

**Comment by pquinn-dls** [comment]
In terms of azimuthal integration there is quite an extensive library developed for x-ray systems, 
pyFAI (https://pyfai.readthedocs.io/en/latest/index.html),  which supports a wider community.  
I'm interested in more x-ray tool integration in hyperspy so a hyperspy-pyFAI libray could cover the radial integration applications for both x-ray and EM applications (also covering differences in physics of the ewald).  There has been some work on pyxem using pyfai which could perhaps be built on.

**Comment by ericpre** [comment]
Just to make the list a bit longer:

https://github.com/CSSFrancis/empyer
https://github.com/py4dstem/py4DSTEM

@CSSFrancis, @bsavitzky

**Comment by uellue** [comment]
Good point, at https://github.com/LiberTEM/LiberTEM/ we have the same concerns about this topic.

LiberTEM does transformations like, for example, radial integration using dot products with dense or sparse matrices. That turns out to be very efficient in most cases since dot product implementations are highly optimized and we only need one back-end for a wide range of transformations. The functions to generate the mask stack only need to be optimized "reasonably well" since they only run once per worker process. For us it would be great to have libraries that generate mask stacks for such transformations.

How is it solved on other relevant packages? Dot product, too, or directly as function that operates on the data?

We'd be very interested to use and co-develop a collection of mask-related utility functions. We already have a collection in https://github.com/LiberTEM/LiberTEM/blob/master/src/libertem/masks.py, which we'd be more than happy to get rid of if equivalent implementations exist, or merge with other code where our part adds value.

**Comment by CSSFrancis** [comment]
I would be willing to help out where I can with this as well! I think it might be a good idea to start by creating a format to define a set of functions that all take a numpy/dask array and some optional mask and can operate on 2-dimensional arrays when possible and then map to apply the function to every part of a higher dimensional data set.

Then again there are some functions that work more efficiently or better given a higher dimensional data sets.  For the example of finding the center of a pattern and fitting an ellipse to a pattern I have looked at different methods that work operating on one two dimensional array, working on the full 4-D data set or even parts of the larger 4-D data set. 

Overall a good start might be to set a standard format for the functions and a centralized location to start collecting the functions.

**Comment by uellue** [comment]
Regarding a standard format for mask-related functions, we used two conventions in LiberTEM. The first, more obvious one is something like

```
def _make_circular_mask(centerX, centerY, imageSizeX, imageSizeY, radius, antialiased=False):
    pass
```

However, more recently we have explored the options to not use the pixel coordinates themselves, but to pass a parameter map of the coordinates. For example,

```
def _make_circular_mask(centerX, centerY, y, x, radius, antialiased=False):
    pass
```

where `y` and `x` contain the y and x coordinates of each pixel. That has two advantages:

1. We can, in principle, work in physical units
2. We support any detector layout and distortions of the projection

For that reason I'd propose to use a convention of passing parameter maps and then calculating masks as function of these parameter maps. Maybe one could think of handling n-dimensional masks  as well.

**Comment by uellue** [comment]
In terms of related libraries:

- https://scikit-image.org/docs/dev/api/skimage.draw.html
- https://pycairo.readthedocs.io/en/latest/index.html

**Comment by CSSFrancis** [comment]
Let me just see if I understand this a bit better.  You're proposing that all of the mask functions should be general enough to deal with pixel values and real space values? 

So for the example of the circular mask....

```
def _make_circular_mask(centerX, centerY, y, x, radius, antialiased=False):
    x1 = x-centerX
    y1 = y-centerY
    return x1*x1+y1*y1<=radius*radius

realXX, realYY = np.ogrid(0:10:.5,0:10:.5)
xx,yy =  np.ogrid(0:20:,0:20:)

real_mask = _make_circular_mask(3.5, 3.5, real_yy, real_xx, 1.5)
pixel_mask = _make_circular_mask(7,7, yy,  xx,3)
```

Where the real and pixel masks should be the same for the case above.

Do I have that correct?
